### PR TITLE
#8055 allow to load 3dtiles tileset via viewer parameters

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR introduces the possibility to include a 3D Tiles layer with the query parameters. The viewer switches to 3D view it it detect a 3D tiles layer among the one included in the query parameter action `CATALOG:ADD_LAYERS_FROM_CATALOGS`

Here the introduced structures:

> The 3D tiles service endpoint does not contain a default property for the name of the layer and it returns only a single record for this reason the name used in the layers array will be used to apply the title to the added 3D Tiles layer:

```json
{
    "type": "CATALOG:ADD_LAYERS_FROM_CATALOGS",
    "layers": ["My 3D Tiles Layer"],
    "sources": [{ "type":"3dtiles", "url":"https://example.com/tileset-pathname/tileset.json" }]
}
```

GET: `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["My 3D Tiles Layer"],"sources":[{"type":"3dtiles","url":"https://example.com/tileset-pathname/tileset.json"}]}]`

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Minor changes to existing features

## Issue

**What is the current behavior?**

#8055

**What is the new behavior?**

It its possible to add 3D tiles layer with the query parameter action `CATALOG:ADD_LAYERS_FROM_CATALOGS`

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] No

### CODE REVIEW

File: docs/developer-guide/map-query-parameters.md [hunk 1]
```diff
@@ -371,3 +371,29 @@ Data of resulting layer can be additionally filtered by passing "CQL_FILTER" int
...
+
+- Use the `url` property to get the layer
+
+```json
+{
+    "type": "CATALOG:ADD_LAYERS_FROM_CATALOGS",
+    "layers": ["https://example.com/tileset-pathname/tileset.json"],
+    "sources": [{ "type":"3dtiles", "url":"https://example.com/tileset-pathname/tileset.json" }]
+}
+```
```
Comment (offtherailz) [inline docs/developer-guide/map-query-parameters.md:385]: I tried this and I had a "aa5" title, but I don't know where it comes from, can you explain?

**Review by offtherailz — CHANGES_REQUESTED** [review]
I noticed that the 3D switch implementation starts from 2D, then commutes to 3D when the data is available. 

This unnecessary rendering is a little ugly and may be confusing. 
I noticed that instead if you put the proper parameters (used for 3D share for coordinates of the camera) this do not happen. 

So I'd suggest to avoid this auto-switch to 3D option based on layer added, and ask explicitly parameters for 3D version of the map.

File: docs/developer-guide/map-query-parameters.md (hunk 1 above)
Comment (allyoucanmap) [reply docs/developer-guide/map-query-parameters.md:383]: New proposed configuration uses the layer entry as title of the new layer, no more multiple approaches

> The 3D tiles service endpoint does not contain a default property for the name of the layer and it returns only a single record for this reason the name used in the layers array will be used to apply the title to the added 3D Tiles layer:


```json
{
    "type": "CATALOG:ADD_LAYERS_FROM_CATALOGS",
    "layers": ["My 3D Tiles Layer"],
    "sources": [{ "type":"3dtiles", "url":"https://example.com/tileset-pathname/tileset.json" }]
}
```

GET: `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["My 3D Tiles Layer"],"sources":[{"type":"3dtiles","url":"https://example.com/tileset-pathname/tileset.json"}]}]`

File: docs/developer-guide/map-query-parameters.md [hunk 2]
```diff
@@ -371,3 +371,29 @@ Data of resulting layer can be additionally filtered by passing "CQL_FILTER" int
...
+
+GET: `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["https://example.com/tileset-pathname/tileset.json"],"sources":[{"type":"3dtiles","url":"https://example.com/tileset-pathname/tileset.json"}]}]`
+
+- Use the `title` property in the source object to assign the name to the layer
```
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply docs/developer-guide/map-query-parameters.md]: With the new configuration the title property is not visible anymore

File: web/client/epics/catalog.js [hunk 3]
```diff
@@ -206,12 +208,22 @@ export default (API) => ({
                         // return one notification for all records that have not been found
                         actions = [recordsNotFound(allRecordsNotFound)];
                     }
+
+                    const layers = results
```
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply web/client/epics/catalog.js:217]: Fixed in review commits, thanks

**Comment by allyoucanmap** [comment]
> I noticed that the 3D switch implementation starts from 2D, then commutes to 3D when the data is available.
> 
> This unnecessary rendering is a little ugly and may be confusing. I noticed that instead if you put the proper parameters (used for 3D share for coordinates of the camera) this do not happen.
> 
> So I'd suggest to avoid this auto-switch to 3D option based on layer added, and ask explicitly parameters for 3D version of the map.

moved the switch at higher level to ensure the map switch to 3D as soon as possible

File: docs/developer-guide/map-query-parameters.md (hunk 1 above)
Comment (offtherailz) [reply docs/developer-guide/map-query-parameters.md:383]: can layers elements be objects? I should follow the same system of other MapStore. Or maybe options can contain more settings (e.g. localized title?).

I should be as much as possible consistent with the rest of the layers.

**Comment by landryb** [comment]
Fwiw i've retested this PR after [64f6f0a](https://github.com/geosolutions-it/MapStore2/pull/9021/commits/64f6f0a6bb7ebedb676cb6806b4ea3cc75c8eb84) and the switch to 3d mode is indeed done straight at viewer loading, much better. The layer title in the TOC is also correctly set from the `layers` parameter:

http://localhost:8081/#/viewer/config?actions=%5B%7B%22type%22:%22CATALOG:ADD_LAYERS_FROM_CATALOGS%22,%22layers%22:%5B%22My%20title%22%5D,%22sources%22:%5B%7B%22type%22:%223dtiles%22,%22url%22:%22http://3d.craig.fr/datasets/Aurillac/3dtiles/tileset.json%22%7D%5D%7D%5D

i've also tested adding two 3dtiles layer at once and it works, layer titles are properly set and match the order of entries in the sources & layers arrays:
http://localhost:8081/#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["Aurillac","Domerat"],"sources":[{"type":"3dtiles","url":"http://3d.craig.fr/datasets/Aurillac/3dtiles/tileset.json"},{"type":"3dtiles","url":"https://3d.craig.fr/datasets/Domerat_bati3d/3dtiles/tileset.json"}]}]

And finally, adding two 3dtiles layers and a wms layer also works, the viewer is still directly loaded in cesium mode from the start:
http://localhost:8081/#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["Aurillac","Domerat","commune"],"sources":[{"type":"3dtiles","url":"http://3d.craig.fr/datasets/Aurillac/3dtiles/tileset.json"},{"type":"3dtiles","url":"https://3d.craig.fr/datasets/Domerat_bati3d/3dtiles/tileset.json"},{"type":"WMS","url":"https://wms.craig.fr/administratif"}]}]

**Comment by landryb** [comment]
> can layers elements be objects? I should follow the same system of other MapStore. Or maybe options can contain more settings (e.g. localized title?).

@offtherailz just curious, what do you have in mind ? setting `layerOptions` ? afaik, so far the layers array can only contain a list of strings but if it's possible to set some layer options that's .. interesting.

**Comment by allyoucanmap** [comment]
> can layers elements be objects? I should follow the same system of other MapStore. Or maybe options can contain more settings (e.g. localized title?).
> 
> I should be as much as possible consistent with the rest of the layers.

The layer element is a string used initially as text in the filter for the [.textSearch](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/catalog.js#L161) method of the catalog and then used as [identifier](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/catalog.js#L173) to match the specific layer in the catalog. The 3D tiles does not need filter or search text because the record is already one so I simply used this information as default title. I don't think an object is really needed for the moment

**Comment by offtherailz** [comment]
> > can layers elements be objects? I should follow the same system of other MapStore. Or maybe options can contain more settings (e.g. localized title?).
> 
> @offtherailz just curious, what do you have in mind ? setting `layerOptions` ? afaik, so far the layers array can only contain a list of strings but if it's possible to set some layer options that's .. interesting.

The idea is to try to see if we can respect as much as possible the same schema and functionalities of the WMS layers so we can expect more or less the same interface and functionalities.

What we implemented now works well and satisfies the requirements we have, I just asked to @allyoucanmap to check this point.

**Comment by allyoucanmap** [comment]
@offtherailz just improved a bit the parsing of the title and added the related doc, see 
432845f

File: docs/developer-guide/map-query-parameters.md [hunk 4]
```diff
@@ -383,3 +383,16 @@ The 3D tiles service endpoint does not contain a default property for the name o
 ```
 
 GET: `#/viewer/config?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["My 3D Tiles Layer"],"sources":[{"type":"3dtiles","url":"https://example.com/tileset-pathname/tileset.json"}]}]`
+
+If the 3D Tiles needs title with different translation is still possible to include the `options` parameter with the desired translation:
```
Comment (offtherailz) [inline docs/developer-guide/map-query-parameters.md:387]:
```suggestion
For the 3D Tiles you can pass also the layer options, to customize the layer. Here and example to localize the title:
```

**Review by offtherailz — APPROVED** [review]

**Comment by allyoucanmap** [comment]
@ElenaGallo please test on dev and let me know if I can backport on branch 2023.01.xx, thanks

**Comment by ElenaGallo** [comment]
@allyoucanmap the rest passed on DEV, @landryb please backport to 2023.01.xx. Thanks

**Comment by allyoucanmap** [comment]
@landryb don't worry, I'll do the backport of this

**Pull request merged**